paddle/fluid/pir/transforms/onednn/matmul_reshape_add_pass.cc

```diff
@@ -0,0 +1,178 @@
+    const auto &full_int_array1 =
+        pat.Op(paddle::dialect::FullIntArrayOp::name(),
+               {{"value", pat.Attr("value")},
+                {"dtype", pat.Attr("dtype")},
+                {"place", pat.Attr("place")}});
+    const auto &reshape1 = pat.Op(paddle::dialect::ReshapeOp::name());
+    reshape1({&pat.Tensor("x"), &full_int_array1()},
+             {&pat.Tensor("reshape_x"), &pat.Tensor("reshape_x_xshape")});
+
+    const auto &matmul = pat.Op(paddle::dialect::MatmulOp::name(),
+                                {{"transpose_x", pat.Attr("trans_x")},
+                                 {"transpose_y", pat.Attr("trans_y")}});
+    pat.Tensor("matmul_out") = matmul(pat.Tensor("reshape_x"), pat.Tensor("w"));
+
+    const auto &full_int_array3 =
+        pat.Op(paddle::dialect::FullIntArrayOp::name());
+    pat.Tensor("shape") = full_int_array3();
+    const auto &reshape3 = pat.Op(paddle::dialect::ReshapeOp::name());
+    reshape3({&pat.Tensor("matmul_out"), &pat.Tensor("shape")},
+             {&pat.Tensor("out"), &pat.Tensor("xshape")});
+
+    const auto &add_ = pat.Op(paddle::dialect::AddOp::name());
+    pat.Tensor("add_out") = as_x_ ? add_(pat.Tensor("out"), pat.Tensor("y"))
+                                  : add_(pat.Tensor("y"), pat.Tensor("out"));
+
+    pat.AddConstraint([&](const paddle::drr::MatchContext &match_ctx) {
+      auto w_dtype = pir::GetDataTypeFromValue(match_ctx.Tensor("w"));
+      if (!w_dtype.isa<pir::Float16Type>() &&
+          !w_dtype.isa<pir::Float32Type>() &&
+          !w_dtype.isa<pir::Float64Type>()) {
+        return false;
+      }
+
+      auto w_dims = pir::GetShapeFromValue(match_ctx.Tensor("w"));
+      auto x_dims = pir::GetShapeFromValue(match_ctx.Tensor("reshape_x"));
+      auto y_dims = pir::GetShapeFromValue(match_ctx.Tensor("y"));
+      auto origin_x_dims = pir::GetShapeFromValue(match_ctx.Tensor("x"));
+
+      if (origin_x_dims.size() < 2) {
+        return false;
+      }
+
+      if (w_dims.size() != 2 || x_dims.size() < 2) {
+        return false;
+      }
+      // Currently，FcOp and GemmEpilogueOp support only RRR format
+      if (x_dims.at(x_dims.size() - 1) != w_dims.at(0) ||
+          match_ctx.Attr<bool>("trans_x") == true ||
+          match_ctx.Attr<bool>("trans_y") == true) {
+        return false;
+      }
+
+      if (y_dims.size() == 1) {
+        return y_dims.at(0) == w_dims.at(1);
+      }
+
+      if (y_dims.size() == 2) {
+        return y_dims.at(0) == 1 && y_dims.at(1) == w_dims.at(1);
+      }
+
+      return false;
+    });
+
+    paddle::drr::ResultPattern res = pat.ResultPattern();
+    const auto &in_num_col_dims_attr =
+        res.ComputeAttr([](const paddle::drr::MatchContext &match_ctx) -> int {
+          auto x_dims = pir::GetShapeFromValue(match_ctx.Tensor("x"));
+          auto reshape_x_dims = match_ctx.Attr<std::vector<int64_t>>("value");
+          int target = reshape_x_dims[0];
```

Comment:
For real model, usually the dim_0 of input is `-1` (cause batch is not decided yet), a.k.a. [-1, m, n, k], and thus reshape could be [-1, m*n*k] or [-1, n*k] or [-1, k]. So instead of directly using `reshape_x_dims[0]`, I recommend refer to [here](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/ir_adaptor/translator/op_translator.cc#L1792) and calculate the attribute.